src/object_gui.cpp

```diff
@@ -25,23 +29,35 @@
 
 #include "safeguards.h"
 
-static ObjectClassID _selected_object_class; ///< the currently visible object class
-static int _selected_object_index;           ///< the index of the selected object in the current class or -1
-static uint8 _selected_object_view;          ///< the view of the selected object
+static const ObjectClass *_selected_object_class;  ///< Currently selected available object class.
+static int _selected_object_index;                 ///< Index of the currently selected object if existing, else \c -1.
+static uint8 _selected_object_view;                ///< the view of the selected object
 
 /** The window used for building objects. */
 class BuildObjectWindow : public Window {
-	static const int OBJECT_MARGIN = 4; ///< The margin (in pixels) around an object.
-	int line_height;                    ///< The height of a single line.
-	int info_height;                    ///< The height of the info box.
-	Scrollbar *vscroll;                 ///< The scrollbar.
+	typedef GUIList<const ObjectClass*, StringFilter&> GUIObjectClassList;
+
+	static const uint EDITBOX_MAX_SIZE = 16;
+	static const int OBJECT_MARGIN = 4;    ///< The margin (in pixels) around an object.
+
+	int line_height;                       ///< The height of a single line.
+	int info_height;                       ///< The height of the info box.
+	Scrollbar* vscroll;                    ///< The scrollbar.
+
+	static Listing   last_sorting;         ///< Default sorting of #GUIObjectClassList.
+	static Filtering last_filtering;       ///< Default filtering of #GUIObjectClassList.
+	static GUIObjectClassList::SortFunction   * const sorter_funcs[]; ///< Sort functions of the #GUIObjectClassList.
+	static GUIObjectClassList::FilterFunction * const filter_funcs[]; ///< Filter functions of the #GUIObjectClassList.
+	GUIObjectClassList object_classes;     ///< Available object classes.
+	StringFilter string_filter;            ///< Filter for available objects.
+	QueryString filter_editbox;            ///< Filter editbox.
 
 	/** Scroll #WID_BO_CLASS_LIST so that the selected object class is visible. */
 	void EnsureSelectedObjectClassIsVisible()
 	{
 		uint pos = 0;
-		for (int i = 0; i < _selected_object_class; i++) {
-			if (ObjectClass::Get((ObjectClassID) i)->GetUISpecCount() == 0) continue;
+		for (uint i = 0; i < object_classes.size(); i++) {
```

Comment:
I think you should use "this->" for object_classes. It will make the PR easier to review.

Reply:
Good point, updated!